driver-core/src/main/com/mongodb/client/model/expressions/Expression.java

```diff
@@ -103,12 +103,13 @@ public interface Expression {
      * @param or
      * @return
      */
-    BooleanExpression isBooleanOr(BooleanExpression or);
-    NumberExpression isNumberOr(NumberExpression or);
-    StringExpression isStringOr(StringExpression or);
-    DateExpression isDateOr(DateExpression or);
-    ArrayExpression<Expression> isArrayOr(ArrayExpression<? extends Expression> or);
-    <T extends DocumentExpression> T isDocumentOr(T or);
+    BooleanExpression isBooleanOr(BooleanExpression other);
+    NumberExpression isNumberOr(NumberExpression other);
+    IntegerExpression isIntegerOr(IntegerExpression other);
+    StringExpression isStringOr(StringExpression other);
+    DateExpression isDateOr(DateExpression other);
+    <T extends Expression> ArrayExpression<T> isArrayOr(ArrayExpression<T> other);
```

Comment:
The parameter type should be `ArrayExpression<? extends T>` for the same reason we agreed on `ArrayExpression<? extends Expression>` in a [previous PR](https://github.com/mongodb/mongo-java-driver/pull/1050/files?show-viewed-files=true&file-filters%5B%5D=#r1030839953), and for the same reason I [requested a change](https://github.com/mongodb/mongo-java-driver/pull/1053#discussion_r1033053835) in a [subsequent PR](https://github.com/mongodb/mongo-java-driver/pull/1053#discussion_r1033053835). The reason is: we want `ArrayExpression<IntegerExpression>` to be a subtype of `ArrayExpression<NumberExpression>` (it's false in Java, but the API may make you feel like it's true).

The code example that does not compile now, but will after the change:
```java
ArrayExpression<NumberExpression> a = ofBooleanArray().isArrayOr(ofIntegerArray());
```